internal/policy/verify.go

```diff
@@ -97,43 +98,118 @@ func VerifyRelativeForRef(ctx context.Context, repo *git.Repository, initialPoli
+				}
+
+				currentPolicy = newPolicy
+				continue
+			}
+
+			if err := verifyEntry(ctx, repo, currentPolicy, entry); err != nil {
+				// If the invalid entry is never marked as skipped, we return err
+				if !entry.SkippedBy(annotations[entry.ID]) {
+					return err
+				}
+
+				// The invalid entry's been marked as skipped but we still need
+				// to see if another entry fixed state for non-gittuf users
+				invalidEntry = entry
+				verificationErr = err
+			}
+			continue
+		}
+
+		// This is only reached when we have an invalid state
+
+		// 1. What's the last good state?
+		// We don't care about its annotations because it's been verified prior
+		// to verifying the current entry, we also know it's unskipped
+		lastGoodEntry, _, err := rsl.GetLatestUnskippedReferenceEntryForRefBefore(repo, invalidEntry.RefName, invalidEntry.ID)
+		if err != nil {
+			return err
+		}
+		lastGoodEntryCommit, err := repo.CommitObject(lastGoodEntry.TargetID)
+		if err != nil {
+			return err
+		}
+		lastGoodTreeID := lastGoodEntryCommit.TreeHash
+
+		// 2. What entries do we have in the current verification set for the
+		// ref? The first one that is tree-same as lastGoodEntry's commit is the
+		// fix. Entries prior to that one in the queue are considered invalid
+		// and must be skipped
+		fixed := false
+		unskippedInvalid := false
+		newEntryQueue := []*rsl.ReferenceEntry{}
+		for len(entries) != 0 {
+			newEntry := entries[0]
+			entries = entries[1:]
+
+			// TODO: this is essentially just refName or policy entries, right?
+			if newEntry.RefName != invalidEntry.RefName {
+				// Unrelated entry that must be processed in the outer loop
+				newEntryQueue = append(newEntryQueue, newEntry)
+				continue
+			}
+
+			newEntryCommit, err := repo.CommitObject(newEntry.TargetID)
 			if err != nil {
 				return err
 			}
+			if newEntryCommit.TreeHash == lastGoodTreeID {
+				// Fix found, we append the rest of the current verification set
+				// to the new entry queue
+				fixed = true
+				newEntryQueue = append(newEntryQueue, entries...)
+				break
+			}
 
-			if err := currentPolicy.VerifyNewState(ctx, newPolicy); err != nil {
-				return err
+			// newEntry is not tree-same / commit-same, so it is automatically
+			// invalid, check that it's been marked as revoked
+			if !newEntry.SkippedBy(annotations[newEntry.ID]) {
+				unskippedInvalid = true
 			}
+		}
 
-			currentPolicy = newPolicy
-			continue
+		if !fixed {
+			// If we haven't found a fix, return the original error
+			return verificationErr
 		}
 
-		if err := verifyEntry(ctx, repo, currentPolicy, entry); err != nil {
-			return err
+		if unskippedInvalid {
+			// We may have found a fix but if an invalid intermediate entry
+			// wasn't skipped, return error
+			return ErrInvalidEntryNotSkipped
 		}
+
+		// Reset these trackers to continue verification with rest of the queue
+		// We may encounter other issues
+		invalidEntry = nil
+		verificationErr = nil
+
+		entries = newEntryQueue
```

Comment:
That's fair. The idea is that we add a) all entries for other refs to the new queue, and b) we append contents of the current queue after a fix to this new queue. So the new queue is composed exclusively of entries that need to be verified still. This is because every reference entry for the affected ref between the first invalid entry and the fix entry are also considered to be invalid and must be skipped. So all of these are excluded from the new queue. What's a good way to feel more secure about this?

> especially given we're applying extra filtering in the lastGoodEntry check

Can you clarify this re the extra filtering? Do you mean the before clause?

Reply:
Resolved off thread - we're appending all other entries about after confirming the sub-entry hasn't been skipped.